File: sdks/go/types/flow.go
package types

import "time"

// FlowStatus type
type FlowStatus string

const (
	FlowStatusActive   FlowStatus = "active"
	FlowStatusInactive FlowStatus = "inactive"
	FlowStatusDraft    FlowStatus = "draft"
)

// FlowExecutionStatus type
type FlowExecutionStatus string

const (
	FlowExecutionStatusRunning   FlowExecutionStatus = "running"
	FlowExecutionStatusWaiting   FlowExecutionStatus = "waiting"
	FlowExecutionStatusCompleted FlowExecutionStatus = "completed"
	FlowExecutionStatusFailed    FlowExecutionStatus = "failed"
	FlowExecutionStatusCancelled FlowExecutionStatus = "cancelled"
)

// FlowNode in a flow
type FlowNode struct {
	ID       string                 `json:"id"`
	Type     string                 `json:"type"`
	Position map[string]float64     `json:"position"`
	Data     map[string]interface{} `json:"data"`
}

// FlowEdge connecting nodes
type FlowEdge struct {
	ID           string `json:"id"`
	Source       string `json:"source"`
	Target       string `json:"target"`
	SourceHandle string `json:"sourceHandle,omitempty"`
	TargetHandle string `json:"targetHandle,omitempty"`
	Label        string `json:"label,omitempty"`
	Condition    string `json:"condition,omitempty"`
}

// FlowVariable in a flow
type FlowVariable struct {
	Name         string      `json:"name"`
	Type         string      `json:"type"`
	DefaultValue interface{} `json:"defaultValue,omitempty"`
	Description  string      `json:"description,omitempty"`
}

// Flow model
type Flow struct {
	ID          string                 `json:"id"`
	TenantID    string                 `json:"tenantId"`
	Name        string                 `json:"name"`
	Description string                 `json:"description,omitempty"`
	Status      FlowStatus             `json:"status"`
	Version     int                    `json:"version"`
	Nodes       []FlowNode             `json:"nodes"`
	Edges       []FlowEdge             `json:"edges"`
	Variables   []FlowVariable         `json:"variables,omitempty"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
	Timestamps
}

// FlowExecutionStep in execution history
type FlowExecutionStep struct {
	NodeID      string                 `json:"nodeId"`
	NodeType    string                 `json:"nodeType"`
	StartedAt   time.Time              `json:"startedAt"`
	CompletedAt *time.Time             `json:"completedAt,omitempty"`
	Input       map[string]interface{} `json:"input,omitempty"`
	Output      map[string]interface{} `json:"output,omitempty"`
	Error       string                 `json:"error,omitempty"`
}

// FlowExecution model
type FlowExecution struct {
	ID             string                 `json:"id"`
	FlowID         string                 `json:"flowId"`
	ConversationID string                 `json:"conversationId"`
	Status         FlowExecutionStatus    `json:"status"`
	CurrentNodeID  string                 `json:"currentNodeId,omitempty"`
	Variables      map[string]interface{} `json:"variables"`
	History        []FlowExecutionStep    `json:"history"`
	StartedAt      time.Time              `json:"startedAt"`
	CompletedAt    *time.Time             `json:"completedAt,omitempty"`
	Error          string                 `json:"error,omitempty"`
}

// CreateFlowInput for creating flows
type CreateFlowInput struct {
	Name        string                 `json:"name"`
	Description string                 `json:"description,omitempty"`
	Nodes       []FlowNode             `json:"nodes,omitempty"`
	Edges       []FlowEdge             `json:"edges,omitempty"`
	Variables   []FlowVariable         `json:"variables,omitempty"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
}

// UpdateFlowInput for updating flows
type UpdateFlowInput struct {
	Name        string                 `json:"name,omitempty"`
	Description string                 `json:"description,omitempty"`
	Status      FlowStatus             `json:"status,omitempty"`
	Nodes       []FlowNode             `json:"nodes,omitempty"`
	Edges       []FlowEdge             `json:"edges,omitempty"`
	Variables   []FlowVariable         `json:"variables,omitempty"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
}

File: sdks/typescript/src/types/flow.ts
/**
 * Flow types
 */

import type { PaginationParams, Timestamps } from './common';

export interface Flow extends Timestamps {
  id: string;
  tenantId: string;
  name: string;
  description?: string;
  status: FlowStatus;
  version: number;
  nodes: FlowNode[];
  edges: FlowEdge[];
  variables?: FlowVariable[];
  metadata?: Record<string, unknown>;
}

export type FlowStatus = 'active' | 'inactive' | 'draft';

export interface FlowNode {
  id: string;
  type: FlowNodeType;
  position: { x: number; y: number };
  data: FlowNodeData;
}

export type FlowNodeType =
  | 'start'
  | 'message'
  | 'condition'
  | 'action'
  | 'input'
  | 'ai'
  | 'api'
  | 'delay'
  | 'assign'
  | 'end';

export interface FlowNodeData {
  label?: string;
  // Message node
  messageContent?: string;
  messageType?: 'text' | 'media' | 'buttons' | 'list';
  buttons?: FlowButton[];
  // Condition node
  conditions?: FlowCondition[];
  // Action node
  actionType?: string;
  actionParams?: Record<string, unknown>;
  // Input node
  inputVariable?: string;
  inputValidation?: InputValidation;
  // AI node
  agentId?: string;
  prompt?: string;
  outputVariable?: string;
  // API node
  apiUrl?: string;
  apiMethod?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  apiHeaders?: Record<string, string>;
  apiBody?: string;
  // Delay node
  delaySeconds?: number;
  // Assign node
  assignments?: VariableAssignment[];
}

export interface FlowButton {
  id: string;
  text: string;
  value?: string;
}

export interface FlowCondition {
  id: string;
  variable: string;
  operator: ConditionOperator;
  value: unknown;
  targetNodeId: string;
}

export type ConditionOperator =
  | 'equals'
  | 'not_equals'
  | 'contains'
  | 'not_contains'
  | 'starts_with'
  | 'ends_with'
  | 'greater_than'
  | 'less_than'
  | 'is_empty'
  | 'is_not_empty'
  | 'matches_regex';

export interface InputValidation {
  type: 'text' | 'number' | 'email' | 'phone' | 'date' | 'regex';
  pattern?: string;
  errorMessage?: string;
  retries?: number;
}

export interface VariableAssignment {
  variable: string;
  value: unknown;
  expression?: string;
}

export interface FlowEdge {
  id: string;
  source: string;
  target: string;
  sourceHandle?: string;
  targetHandle?: string;
  label?: string;
  condition?: string;
}

export interface FlowVariable {
  name: string;
  type: 'string' | 'number' | 'boolean' | 'array' | 'object';
  defaultValue?: unknown;
  description?: string;
}

// Execution types
export interface FlowExecution {
  id: string;
  flowId: string;
  conversationId: string;
  status: FlowExecutionStatus;
  currentNodeId?: string;
  variables: Record<string, unknown>;
  history: FlowExecutionStep[];
  startedAt: string;
  completedAt?: string;
  error?: string;
}

export type FlowExecutionStatus = 'running' | 'waiting' | 'completed' | 'failed' | 'cancelled';

export interface FlowExecutionStep {
  nodeId: string;
  nodeType: FlowNodeType;
  startedAt: string;
  completedAt?: string;
  input?: Record<string, unknown>;
  output?: Record<string, unknown>;
  error?: string;
}

// Request types
export interface CreateFlowInput {
  name: string;
  description?: string;
  nodes?: FlowNode[];
  edges?: FlowEdge[];
  variables?: FlowVariable[];
  metadata?: Record<string, unknown>;
}

export interface UpdateFlowInput {
  name?: string;
  description?: string;
  status?: FlowStatus;
  nodes?: FlowNode[];
  edges?: FlowEdge[];
  variables?: FlowVariable[];
  metadata?: Record<string, unknown>;
}

export interface ListFlowsParams extends PaginationParams {
  status?: FlowStatus;
  search?: string;
}

export interface ExecuteFlowInput {
  conversationId: string;
  variables?: Record<string, unknown>;
  startNodeId?: string;
}
